### PR TITLE
ReactDOM.useEvent: fix scope propagation issue

### PR DESCRIPTION
**Pull request by trueadm**
This PR fixes an issue with `useEvent` and scopes (see attached test). Specifically, we need to pass through the instance of the scope, so we can properly track if we can stopPropagation. Previously we used the `currentTarget`, which is the last physical DOM element. In the cases of scopes, this can be problematic if we have:

```jsx
<Scope>
  <Scope>
    <div />
  </Scope>
</Scope>
```

To get around this we instead pass around an object, as `{ instance: Fiber | null, container: EventTarget }`. We use this object to maintain the instance and contain (currentTarget) separately. I also moved the logic for this into the modern event system module so we don't affect other event systems with this change.

### CODE REVIEW

File: packages/legacy-events/EventPluginUtils.js [hunk 1]
```diff
@@ -80,20 +80,12 @@ export function executeDispatchesInOrder(event) {
     validateEventDispatches(event);
   }
   if (Array.isArray(dispatchListeners)) {
-    let previousInstance;
     for (let i = 0; i < dispatchListeners.length; i++) {
-      const instance = dispatchInstances[i];
```
Comment (gaearon) [inline packages/legacy-events/EventPluginUtils.js:85]: Can you explain what this code was doing? I'm not sure I understand why we would have the same instance twice, or how that relates to propagation. I see it was added very recently? Is this used by the legacy event system? Seems a bit scary to change it.

Comment (gaearon) [reply packages/legacy-events/EventPluginUtils.js:85]: I think this PR is essentially removing the modified code, and instead forking it for modern system only?

File: packages/react-dom/src/events/DOMModernPluginEventSystem.js [hunk 2]
```diff
@@ -165,6 +167,57 @@ export const reactScopeListenerStore: WeakMap<
...
+      // - as an object { instance: Fiber | null, container: EventTarget }
+      // The second form is used for useEvent types
+      if (dispatchInstance.container !== undefined) {
+        instance = ((dispatchInstance: any): DispatchInstance).instance;
+        currentTarget = ((dispatchInstance: any): DispatchInstance).container;
+      } else {
+        currentTarget = getNodeFromInstance(((dispatchInstance: any): Fiber));
+      }
+      // We check if the instance was the same as the last one,
+      // if it was, then we're still on the same instance thus
```
Comment (gaearon) [inline packages/react-dom/src/events/DOMModernPluginEventSystem.js:205]: What does it mean to be on the same instance? Several `useEvent` targeting the same node?

File: packages/legacy-events/EventPluginUtils.js (hunk 1 above)
Comment (trueadm) [reply packages/legacy-events/EventPluginUtils.js:85]: This is reverting this file back to what it was before any of the changes from the modern event system were made to it. Like you say, we instead use a forked version instead.

File: packages/react-dom/src/events/DOMModernPluginEventSystem.js [hunk 3]
```diff
@@ -165,6 +167,57 @@ export const reactScopeListenerStore: WeakMap<
...
+  let previousInstance;
+
+  if (dispatchListeners !== null && dispatchInstances !== null) {
+    for (let i = 0; i < dispatchListeners.length; i++) {
+      const dispatchInstance = dispatchInstances[i];
+      const dispatchListener = dispatchListeners[i];
+      let instance = dispatchInstance;
+      let currentTarget;
+
+      // dispatchInstance can come in two forms:
```
Comment (gaearon) [inline packages/react-dom/src/events/DOMModernPluginEventSystem.js:194]: I find this confusing. I worry these special cases are making modern event system harder to follow than the legacy one. Imagine we add a `container` field to Fiber later and this will brek.

If we really need to keep more data here, can we always wrap them in objects?

File: packages/react-dom/src/events/DOMModernPluginEventSystem.js (hunk 2 above)
Comment (trueadm) [reply packages/react-dom/src/events/DOMModernPluginEventSystem.js:205]: Exactly, it's for `useEvent` targeting the same node.

File: packages/react-dom/src/events/DOMModernPluginEventSystem.js (hunk 3 above)
Comment (trueadm) [reply packages/react-dom/src/events/DOMModernPluginEventSystem.js:194]: Yeah, I see that. I'll revise.

Comment (gaearon) [reply packages/react-dom/src/events/DOMModernPluginEventSystem.js:194]: Or I guess a new field so that we don't put more things into `_dispatchInstances`. Like `_dispatchContainers`.

File: packages/react-dom/src/events/DOMModernPluginEventSystem.js [hunk 4]
```diff
@@ -165,6 +167,57 @@ export const reactScopeListenerStore: WeakMap<
...
+        currentTarget = ((dispatchInstance: any): DispatchInstance).container;
+      } else {
+        currentTarget = getNodeFromInstance(((dispatchInstance: any): Fiber));
+      }
+      // We check if the instance was the same as the last one,
+      // if it was, then we're still on the same instance thus
+      // propagation should not stop. If we add support for
+      // stopImmediatePropagation at some point, then we'll
+      // need to handle that case here differently.
+      if (instance !== previousInstance && event.isPropagationStopped()) {
```
Comment (gaearon) [inline packages/react-dom/src/events/DOMModernPluginEventSystem.js:209]: I'm still not sure I understand this piece of logic. We check for propagation for the first `useEvent` OK. Then we have another `useEvent` on the same node. Let's say this one calls `stopPropagation`. Would we ignore it? Why?

File: packages/react-dom/src/events/accumulateTwoPhaseListeners.js [hunk 5]
```diff
@@ -111,18 +111,19 @@ export default function accumulateTwoPhaseListeners(
...
           const captureListeners = Array.from(listeners.captured);
           const bubbleListeners = Array.from(listeners.bubbled);
+          const container = ((lastHostComponent: any): Element);
 
           for (let i = 0; i < captureListeners.length; i++) {
             const listener = captureListeners[i];
             const {callback} = listener;
             dispatchListeners.unshift(callback);
-            dispatchInstances.unshift(((lastHostComponent: any): Element));
+            dispatchInstances.unshift({instance: node, container});
```
Comment (gaearon) [inline packages/react-dom/src/events/accumulateTwoPhaseListeners.js:120]: Can you explain in more detail what's different about Scopes that makes us need this? Why do we need to "remember" the closest "last host component", but we don't for regular events? What is special here? Why can't we figure out the DOM node later?

Comment (trueadm) [reply packages/react-dom/src/events/accumulateTwoPhaseListeners.js:120]: Scopes don't have a physical DOM element representing them as a target, so in order to provide the user with one, we given them the nearest DOM node from the
the native event's original target – this is also how React Flare worked. We can't do this later, as we don't know which part of the tree we came in by – so we have to do this here.

File: packages/react-dom/src/events/DOMModernPluginEventSystem.js (hunk 4 above)
Comment (trueadm) [reply packages/react-dom/src/events/DOMModernPluginEventSystem.js:209]: We'd ignore it as `stopPropagation` works like that on the DOM today. We'd need to add `stopImmediatePropagation` that works like how you probably want, where we can stop propagation even on the same instance.

File: packages/react-dom/src/events/accumulateTwoPhaseListeners.js (hunk 5 above)
Comment (gaearon) [reply packages/react-dom/src/events/accumulateTwoPhaseListeners.js:120]: Is this about `target` or `currentTarget`?

File: packages/react-dom/src/events/DOMModernPluginEventSystem.js (hunk 3 above)
Comment (trueadm) [reply packages/react-dom/src/events/DOMModernPluginEventSystem.js:194]: I was doing just that :)

File: packages/react-dom/src/events/accumulateTwoPhaseListeners.js (hunk 5 above)
Comment (trueadm) [reply packages/react-dom/src/events/accumulateTwoPhaseListeners.js]: It's about the `currentTarget` on the event object for when events are on scopes. We want it to be the last "real" currentTarget that is in the DOM.

File: packages/react-dom/src/events/DOMModernPluginEventSystem.js (hunk 4 above)
Comment (gaearon) [reply packages/react-dom/src/events/DOMModernPluginEventSystem.js:201]: Hmm. I'm not sure I understand what you mean by DOM working like this.

Can you provide an example?

>We'd need to add stopImmediatePropagation that works like how you probably want

I am not referring to one `useEvent` preventing another `useEvent`.

I am talking about the second `useEvent` being able to prevent *further* propagation up the tree. Does that work?

Comment (trueadm) [reply packages/react-dom/src/events/DOMModernPluginEventSystem.js:201]: > I am talking about the second useEvent being able to prevent further propagation up the tree. Does that work?

Yes it does, by design.

> Hmm. I'm not sure I understand what you mean by DOM working like this.

I mean that it works in same way propagation works on the DOM. We move a pointer up and down a single-list (which is section of the node tree) and each level of the tree is a propagation layer. If something stops propagation, we don't exit processing of the list until we move the pointer to the next level.

Comment (gaearon) [reply packages/react-dom/src/events/DOMModernPluginEventSystem.js:201]: >Yes it does, by design.

Do you mind adding a test for this?

Comment (trueadm) [reply packages/react-dom/src/events/DOMModernPluginEventSystem.js:201]: That's what the test on this PR should be testing, unless we're talking about different things.

Comment (gaearon) [reply packages/react-dom/src/events/DOMModernPluginEventSystem.js:201]: OK I get it now. What you're saying is that by the time we reach the next node, we will run this check again, and so if it has been stopped by then, we won't dispatch for it.

File: packages/react-dom/src/events/accumulateEnterLeaveListeners.js [hunk 6]
```diff
@@ -83,24 +83,26 @@ function accumulateEnterLeaveListenersForEvent(
...
     if (node === common) {
       break;
     }
-    const alternate = node.alternate;
+    const {alternate, stateNode, tag} = node;
     if (alternate !== null && alternate === common) {
       break;
     }
-    if (node.tag === HostComponent) {
+    if (stateNode !== null && tag === HostComponent) {
```
Comment (gaearon) [inline packages/react-dom/src/events/accumulateEnterLeaveListeners.js:97]: Is this a behavior change?

Either it was never null. And in that case it's safe to use it.

Or it was sometimes null. And then excluding this component from traversal changes the behavior without any flags.

File: packages/legacy-events/SyntheticEvent.js [hunk 7]
```diff
@@ -78,6 +78,7 @@ function SyntheticEvent(
   this.nativeEvent = nativeEvent;
   this._dispatchListeners = null;
   this._dispatchInstances = null;
+  this._dispatchContainers = null;
```
Comment (gaearon) [inline packages/legacy-events/SyntheticEvent.js:81]: This also needs to be cleared in `destructor`.  Otherwise we're risking a memory leak.

File: packages/react-dom/src/events/accumulateEnterLeaveListeners.js (hunk 6 above)
Comment (trueadm) [reply packages/react-dom/src/events/accumulateEnterLeaveListeners.js:97]: If we the stateNode is `null` it saves us processing the listener for the node as it's unmounted and we can avoid CPU cycles.

Comment (gaearon) [reply packages/react-dom/src/events/accumulateEnterLeaveListeners.js:97]: What happens today if the node is unmounted in the middle of an event?

File: packages/react-dom/src/events/accumulateEventTargetListeners.js [hunk 8]
```diff
@@ -32,7 +34,8 @@ export default function accumulateEventTargetListeners(
           const listener = captureListeners[i];
           const {callback} = listener;
           dispatchListeners.push(callback);
-          dispatchInstances.push(targetContainer);
+          dispatchInstances.push(null);
```
Comment (gaearon) [inline packages/react-dom/src/events/accumulateEventTargetListeners.js:37]: Why are these null?

File: packages/react-dom/src/events/accumulateTwoPhaseListeners.js [hunk 9]
```diff
@@ -67,9 +68,11 @@ export default function accumulateTwoPhaseListeners(
               if (capture === true) {
                 dispatchListeners.unshift(callback);
                 dispatchInstances.unshift(node);
+                dispatchContainers.unshift(instance);
```
Comment (gaearon) [inline packages/react-dom/src/events/accumulateTwoPhaseListeners.js:71]: The naming is really tripping me. We're pushing a "node" to "instances" but an "instance" to "containers". What is the mental model for what's actually in there? Can we rename these fields and variable to match, and clarify which are DOM nodes and which are Fibers and which are something else?

Comment (gaearon) [reply packages/react-dom/src/events/accumulateTwoPhaseListeners.js:71]: I don't mean to be pedantic. I find it very difficult to wrap my mind around this, especially because we have several cases. Right now I don't even have the confidence any particular field always holds things of the same type.

File: packages/react-dom/src/events/accumulateEnterLeaveListeners.js (hunk 6 above)
Comment (trueadm) [reply packages/react-dom/src/events/accumulateEnterLeaveListeners.js:97]: If it were `null` today, we'd invoke `invokeGuardedCallbackAndCatchFirstError` with a `null` `currentTarget` on the synthetic event.

File: packages/react-dom/src/events/accumulateEventTargetListeners.js (hunk 8 above)
Comment (trueadm) [reply packages/react-dom/src/events/accumulateEventTargetListeners.js:37]: We don't have an instance for `EventTarget` listeners, because they don't need or might even have a backing fiber (i.e. `window` or `document`).

File: packages/react-dom/src/events/accumulateEnterLeaveListeners.js (hunk 6 above)
Comment (gaearon) [reply packages/react-dom/src/events/accumulateEnterLeaveListeners.js:97]: If this is purely optimization I'd prefer to leave it out. Seems like unnecessary risk if this code is shared by the legacy system unless we have a clear proof that this does not change observable behavior.

Comment (gaearon) [reply packages/react-dom/src/events/accumulateEnterLeaveListeners.js:97]: >we'd invoke invokeGuardedCallbackAndCatchFirstError with a null currentTarget on the synthetic event.

That seems like observably different behavior? User code would get called.

File: packages/react-dom/src/events/accumulateTwoPhaseListeners.js (hunk 9 above)
Comment (trueadm) [reply packages/react-dom/src/events/accumulateTwoPhaseListeners.js:71]: Valid point, I'll clear up the names.

Comment (gaearon) [reply packages/react-dom/src/events/accumulateTwoPhaseListeners.js:71]: How about this:

```
_dispatchListeners
_dispatchFibers
_dispatchCurrentTargets
```

?

File: packages/react-dom/src/events/accumulateEnterLeaveListeners.js (hunk 6 above)
Comment (trueadm) [reply packages/react-dom/src/events/accumulateEnterLeaveListeners.js:97]: Actually scrap what I said above, we call `getListener` later and that again checks for `stateNode` and if it's `null` then we don't fire any user code. So this change should be fine and have no obserable behavior changes.

File: packages/react-dom/src/events/accumulateTwoPhaseListeners.js (hunk 9 above)
Comment (trueadm) [reply packages/react-dom/src/events/accumulateTwoPhaseListeners.js]: I don't want to change from `dispatchInstances` as that's what the existing legacy system uses and changing that is far too risky. I can change the `currentContainers` to `_dispatchCurrentTargets` though.

File: packages/react-dom/src/events/accumulateTwoPhaseListeners.js [hunk 10]
```diff
@@ -31,20 +31,22 @@ export default function accumulateTwoPhaseListeners(
...
 
   // Accumulate all instances and listeners via the target -> root path.
-  while (node !== null) {
-    const {stateNode: instance, tag} = node;
+  while (instance !== null) {
+    const {stateNode, tag} = instance;
     // Handle listeners that are on HostComponents (i.e. <div>)
-    if (instance !== null && tag === HostComponent) {
-      lastHostComponent = instance;
+    if (stateNode !== null && tag === HostComponent) {
```
Comment (gaearon) [inline packages/react-dom/src/events/accumulateTwoPhaseListeners.js:47]: Let's make tag checks first here and elsewhere. We don't care about other state nodes anyway.

File: packages/legacy-events/ReactSyntheticEventType.js [hunk 11]
```diff
@@ -40,8 +40,11 @@ export type ReactSyntheticEvent = {|
     nativeEventTarget: EventTarget,
   ) => ReactSyntheticEvent,
   isPersistent: () => boolean,
-  _dispatchInstances: null | Array<Fiber | EventTarget> | Fiber | EventTarget,
-  _dispatchListeners: null | Array<Function> | Function,
+  isPropagationStopped: () => boolean,
+  _dispatchInstances: null | Array<Fiber | null>,
```
Comment (gaearon) [inline packages/legacy-events/ReactSyntheticEventType.js:44]: OK, we're back to these being strict. This is a lot clearer IMO.

Now can we revert the changes to `executeDispatch` in `EventPluginUtils` from https://github.com/facebook/react/commit/a600408b284e92ed3d801d7f4e37d10b5da06b10?

I suggest:

```diff
- event.currentTarget =
-    inst.tag !== undefined ? getNodeFromInstance(inst) : inst;
+ event.currentTarget = getNodeFromInstance(inst);
```

I guess this isn't even used by the modern code path anymore.

Comment (gaearon) [reply packages/legacy-events/ReactSyntheticEventType.js:44]: When did it stop being possible to have a single `Fiber` in there? Is it still possible in legacy system?

Comment (trueadm) [reply packages/legacy-events/ReactSyntheticEventType.js:44]: In the legacy system it is, but not in the modern system. Given we've forked it, the old code should continue to work. I've re-added the type anyway, even if it's going away very soon.
